Add IOperation test helper to verify operation tree and compilation d…

mavasani opened the pull request:
…iagnostics

mavasani commented:
Tagging @AlekseyTs @dotnet/analyzer-ioperation for review. Example generated tests:

VB:
```
<Fact()>
        Public Sub XXXX()
            Dim source = <![CDATA[
Imports System
Imports System.Collections.Generic
Imports System.Linq

Module Program
    Sub Main(args As String())
        Console.WriteLine(a)'BIND:"Console.WriteLine(a)"
        Console.WriteLine(b)
    End Sub
End Module]]>.Value

            Dim expectedOperationTree = <![CDATA[
IExpressionStatement (OperationKind.ExpressionStatement, IsInvalid)
  IInvalidExpression (OperationKind.InvalidExpression, Type: System.Void, IsInvalid)
]]>.Value

            ' BC30451: 'a' is not declared. It may be inaccessible due to its protection level.
            ' BC30451: 'b' is not declared. It may be inaccessible due to its protection level.
            Dim expectedDiagnostics = New DiagnosticDescription() {
                Diagnostic(ERRID.ERR_NameNotDeclared1, "a").WithArguments("a").WithLocation(8, 27),
                Diagnostic(ERRID.ERR_NameNotDeclared1, "b").WithArguments("b").WithLocation(9, 27)
            }

            VerifyOperationTreeAndDiagnosticsForTest(Of ExpressionStatementSyntax)(source, expectedOperationTree, expectedDiagnostics)
        End Sub
```

C#:
```
[Fact]
        public void XXXX2()
        {
            string source = @"
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

class Program
{
    static void Main(string[] args)
    {
        /*<bind>*/Console.WriteLine(a);/*</bind>*/
        Console.WriteLine(b);
    }
}
";
            string expectedOperationTree = @"
IExpressionStatement (OperationKind.ExpressionStatement, IsInvalid)
  IInvocationExpression ( void System.Console.WriteLine()) (OperationKind.InvocationExpression, Type: System.Void, IsInvalid)
    Instance Receiver: IOperation:  (OperationKind.None)
";
            var expectedDiagnostics = new DiagnosticDescription[] {
                // CS0103: The name 'a' does not exist in the current context
                Diagnostic(ErrorCode.ERR_NameNotInContext, "a").WithArguments("a").WithLocation(11, 37),
                // CS0103: The name 'b' does not exist in the current context
                Diagnostic(ErrorCode.ERR_NameNotInContext, "b").WithArguments("b").WithLocation(12, 27)
            };

            VerifyOperationTreeAndDiagnosticsForTest<ExpressionStatementSyntax>(source, expectedOperationTree, expectedDiagnostics);
        }
```

genlu commented on src/Compilers/Test/Utilities/CSharp/SemanticModelTestBase.cs:
```diff
@@ -2,6 +2,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Collections.Immutable;
```
Is this import necessary?

AlekseyTs commented:
Please change VB to use XML as the baseline similar to AssertTheseDiagnostics helper.

For C#, please include source line into the comment within the base line:
```
                // CS0103: The name 'a' does not exist in the current context
               // ---------------source line goes here---------------
                Diagnostic(ErrorCode.ERR_NameNotInContext, "a").WithArguments("a").WithLocation(11, 37),
```
Similar to how VerifyDiaggnostics helper dumps the actual set in case of an error.

mavasani commented:
@AlekseyTs Does this look fine?

mavasani commented:
Oh I didn't see your prior comment, let me check that

mavasani commented:
@AlekseyTs - I have addressed the feedback.

VB:
```
        <Fact()>
        Public Sub XXXX()
            Dim source = <![CDATA[
Imports System
Imports System.Collections.Generic
Imports System.Linq

Module Program
    Sub Main(args As String())
        Console.Writ'BIND:"Console.Writ"
eLine(a)
        Console.WriteLine(b)
    End Sub
End Module]]>.Value

            Dim expectedOperationTree = <![CDATA[
IExpressionStatement (OperationKind.ExpressionStatement, IsInvalid)
  IInvalidExpression (OperationKind.InvalidExpression, Type: ?, IsInvalid)
]]>.Value

            Dim expectedDiagnostics = <![CDATA[
BC30456: 'Writ' is not a member of 'Console'.
        Console.Writ'BIND:"Console.Writ"
        ~~~~~~~~~~~~
BC30451: 'eLine' is not declared. It may be inaccessible due to its protection level.
eLine(a)
~~~~~
BC30451: 'a' is not declared. It may be inaccessible due to its protection level.
eLine(a)
      ~
BC30451: 'b' is not declared. It may be inaccessible due to its protection level.
        Console.WriteLine(b)
                          ~
]]>.Value

            VerifyOperationTreeAndDiagnosticsForTest(Of ExpressionStatementSyntax)(source, expectedOperationTree, expectedDiagnostics)
        End Sub
```

C#:
```
        [Fact]
        public void XXXX()
        {
            string source = @"
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

class Program
{
    static void Main(string[] args)
    {
        /*<bind>*/Console.Write
/*</bind>*/Line(a);
        Console.WriteLine(b);
    }
}
";
            string expectedOperationTree = @"
IVariableDeclarationStatement (1 variables) (OperationKind.VariableDeclarationStatement)
  IVariableDeclaration: System.Console.Write  (OperationKind.VariableDeclaration)
";
            var expectedDiagnostics = new DiagnosticDescription[] {
                // CS1001: Identifier expected
                //         /*<bind>*/Console.Write
                Diagnostic(ErrorCode.ERR_IdentifierExpected, "").WithLocation(11, 32),
                // CS1002: ; expected
                //         /*<bind>*/Console.Write
                Diagnostic(ErrorCode.ERR_SemicolonExpected, "").WithLocation(11, 32),
                // CS0426: The type name 'Write' does not exist in the type 'Console'
                //         /*<bind>*/Console.Write
                Diagnostic(ErrorCode.ERR_DottedTypeNameNotFoundInAgg, "Write").WithArguments("Write", "System.Console").WithLocation(11, 27),
                // CS0103: The name 'Line' does not exist in the current context
                // /*</bind>*/Line(a);
                Diagnostic(ErrorCode.ERR_NameNotInContext, "Line").WithArguments("Line").WithLocation(12, 12),
                // CS0103: The name 'a' does not exist in the current context
                // /*</bind>*/Line(a);
                Diagnostic(ErrorCode.ERR_NameNotInContext, "a").WithArguments("a").WithLocation(12, 17),
                // CS0103: The name 'b' does not exist in the current context
                //         Console.WriteLine(b);
                Diagnostic(ErrorCode.ERR_NameNotInContext, "b").WithArguments("b").WithLocation(13, 27)
            };

            VerifyOperationTreeAndDiagnosticsForTest<LocalDeclarationStatementSyntax>(source, expectedOperationTree, expectedDiagnostics);
        }
```

AlekseyTs commented:
Could you update how the example generated tests would look like?

mavasani commented:
@AlekseyTs - yes I added them here: https://github.com/dotnet/roslyn/pull/18338#issuecomment-290835090

mavasani commented:
@dotnet-bot retest windows_debug_vs-integration_prtest  please
